### DISCUSSION TITLE
Send pre-encoded data from GStreamer via aiortc

### DISCUSSION THREAD

**Discussion by mfoerst** [aiortc/aiortc#769] [Q&A]
Hi everybody,
I would like to send a video stream from a v4l2-camera to a browser using aiortc. I used the `webcam.py` example and it worked out of the box.

However, for my application I need to do some pre-processing (cropping, rotating, debayering, ...) on the video. I want to use GStreamer for the preprocessing and also for the encoding (H264) since it offers a greater flexibility and more options than pyAV can do and it supports the hardware accelerated encoding on my target embedded device.

Out of the GStreamer pipeline I get H264-encoded RTP packets that must be sent to the receiving browser by aiortc.
I think the correct way to do this, is to create a subclass of `MediaStreamTrack` that hands over the packets to the `RTCRtpSender` via the `recv()` function.
Here I have the problem that `RTCRtpSender` expects `recv()` to return `av.Frame` or `av.Packet` I found no solution to create a `Packet`-object just out of the bytearray that I get from GStreamer.

My solution is to extend the aiortc code to be able to handle packets directly from byte-arrays:
I created an abstract subclass of `MediaStreamTrack` with a modified `recv`-function that return the a list of bytearrays and the corresponding frame timestamp:
```
class EncodedMediaStreamTrack(MediaStreamTrack):
    @abstractmethod
    async def recv(self) -> Tuple[List[bytes], int]:
        pass
```

To handle the new return type, I extended  `_next_encoded_frame` in `RTCRtpSender`:
```
async def _next_encoded_frame(self, codec: RTCRtpCodecParameters):
    if isinstance(self.__track, EncodedMediaStreamTrack):
        # Frames come already encoded from the MediaStreamTrack
        payloads, timestamp = await self.__track.recv()
        audio_level = None

    else:
        # get frame
        frame = await self.__track.recv()
        audio_level = None
        if isinstance(frame, AudioFrame):
            audio_level = rtp.compute_audio_level_dbov(frame)

        # encode frame
        if self.__encoder is None:
            self.__encoder = get_encoder(codec)
        force_keyframe = self.__force_keyframe
        self.__force_keyframe = False
        payloads, timestamp = await self.__loop.run_in_executor(
            None, self.__encoder.encode, frame, force_keyframe
        )

    return RTCEncodedFrame(payloads, timestamp, audio_level)
```

This solution works well, but it requires the source code modification of aiortc.

Does anybody of you know another way to send the data from GStreamer via aiortc?
Could the option to use self-encoded data without pyAV also be useful for other users and is there a chance to merge it into aiortc?

**Comment by unitree-czk** [comment]
That's what I want. If some one could propose some idea, that would be great!
Sometimes the video processing is done elsewhere, and I need someting to bind different videos into one webrtc connection.

**Comment by dev-andrew-smith** [comment]
To me this looks like a great idea, do you have a basic working example? Like videotestsrc -> encoding -> mediastream -> aiortc 
So basically like the server or webcam example
does your gstreamer pipeline use appsink at the end? and if so how are those buffers sent to aiortc

**Comment by antopio26** [comment]
Can this modification be integrated inside the library? I need to send the same frames to multiple clients, with the current implementation of the media track the same video gets encoded as many times as the connected clients, while I would just need it to encode once and send the same packets to all of them.

**Reply by neilyoung** [reply]
A bit late, but you could use a MediaServer for the one-to-many distribution, e.g. Kurento or MediaSoup.

**Comment by elephantum** [comment]
I came here to ask this question, but it seems, that there's a solution already.

I think this functionality is generally useful for embedded use-case, in last several years the way to do hw-accelerated h264 on Raspberry changed at least two times. Gstreamer is usually supported through these changes.

Maybe there's a way to upstream this or similar approach to aiortc?
I think I will be able to make a PR
